app/src/main/java/com/github/polybooks/core/database/implementation/OLBookDatabase.kt

```diff
@@ -0,0 +1,259 @@
+
+import android.annotation.SuppressLint
+import android.os.Build
+import androidx.annotation.RequiresApi
+import com.github.polybooks.core.Book
+import com.github.polybooks.core.Interest
+import com.github.polybooks.core.database.DatabaseException
+import com.github.polybooks.core.database.interfaces.BookDatabase
+import com.github.polybooks.core.database.interfaces.BookOrdering
+import com.github.polybooks.core.database.interfaces.BookOrdering.*
+import com.github.polybooks.core.database.interfaces.BookQuery
+import com.google.firebase.Timestamp
+import com.google.gson.JsonArray
+import com.google.gson.JsonElement
+import com.google.gson.JsonObject
+import java.io.FileNotFoundException
+import java.text.SimpleDateFormat
+import java.util.*
+import java.util.concurrent.CompletableFuture
+import java.util.concurrent.CompletionException
+
+
+private const val TITLE_FIELD_NAME = "full_title"
+private const val AUTHORS_FIELD_NAME = "authors"
+private const val FORMAT_FIELD_NAME = "physical_format"
+private const val ISBN13_FIELD_NAME = "isbn_13"
+private const val PUBLISHER_FIELD_NAME = "publishers"
+private const val PUBLISH_DATE_FIELD_NAME = "publish_date"
+private const val AUTHOR_NAME_FIELD_NAME = "name"
+
+private const val DATE_FORMAT = "MMM dd, yyyy"
+private const val ISBN13_FORMAT = """[0-9]{13}"""
+private const val ISBN10_FORMAT = """[0-9]{9}[0-9X]"""
+private const val ISBN_FORMAT = """($ISBN10_FORMAT)|($ISBN13_FORMAT)"""
+
+private const val OL_BASE_ADDR = """https://openlibrary.org"""
+
+/**
+ * An implementation of a book database based on the Open Library online database
+ * */
+class OLBookDatabase(private val url2json : (String) -> CompletableFuture<JsonElement>) : BookDatabase {
+
+    override fun queryBooks(): BookQuery = OLBookQuery()
+
+    private inner class OLBookQuery : BookQuery {
+
+        private var ordering = DEFAULT
+
+        private var empty: Boolean = true
+        private var title: String? = null
+        private var isbn: String? = null
+
+        override fun onlyIncludeInterests(interests: Collection<Interest>): BookQuery {
+            System.err.println("Warning: onlyIncludeInterest not fully implemented for OLBookQuery")
+            this.empty = true
+            return this
+        }
+
+        override fun searchByTitle(title: String): BookQuery {
+            System.err.println("Warning: search by title not fully implemented for OLBookQuery")
+            this.empty = true
+            return this
+        }
+
+        override fun searchByISBN13(isbn13: String): BookQuery {
+            val regularised = regulariseISBN(isbn13) ?: throw IllegalArgumentException("Given ISBN is not valid")
+            this.empty = false
+            this.title = null
+            this.isbn = regularised
+            return this
+        }
+
+        override fun withOrdering(ordering: BookOrdering): BookQuery {
+            this.ordering = ordering
+            return this
+        }
```

Comment:
Simple question : Is it working now? I mean, will `getAll()` return the book ordered now?

Reply:
getAll returns at most one book for a OLBookDB, so there is not much to order^^